Check if inputs have changed. Fixes #1.

hildjj opened the pull request:
- Tweaked README.md language about JSON file in
the notes at the end.
- Tweaked index.js output so that prettier should
not need to modify it.

Depends on PR #3.

hildjj committed:
Added javascript export.  Fixes #2.

- build.js now outputs files directly, rather than
  needing to have its stdout redirected.
- build.js now outputs two files:
  - index.json for backward compatibility
  - index.js for consistent importing
- index.js has a default export that matches the
  .json file
- index.js has a few other exports, for tooling usage:
  - lastModified: the string the IANA http server
    returned in the 'last-modified' header.
  - updated: the time the build.js script was run.
  - hash: a SHA256 hash of the JSON string.
- Updated README.md with new import instructions.

hildjj committed:
Check if inputs have changed.  Fixes #1.

- Tweaked README.md language about JSON file in
  the notes at the end.
- Tweaked index.js output so that prettier should
  not need to modify it.

Depends on PR #3.

AsherJingkongChen commented:
Hi, I am not sure why this is not a commit in #3. It duplicates #3 at some parts. May you close #3?

AsherJingkongChen commented:
Sorry, I also need time to investigate it, especially how it differs from the original version.

hildjj commented:
This builds on the changes in #3.  Happy to close that in favor of this instead of splitting it up.

hildjj committed:
Fix bullet numbering

AsherJingkongChen commented:
Is there any case that we need `index.js` meanwhile `index.json` exists? Oh wait it preserves backward compatibility and adds caching tags.

AsherJingkongChen commented on build.js:
```diff
@@ -1,16 +1,39 @@
     const { ok, status, statusText } = response;
+    if (status === 304) {
+      console.error(statusText);
+      process.exit(2);
```
Nice early exiting.

AsherJingkongChen commented on build.js:
```diff
@@ -40,8 +63,34 @@
+      `\
+// @generated by build.js
+export const updated = new Date('${new Date().toISOString()}');
+export const lastModified = '${response.headers.get('last-modified')}';
```
Not safe, should escape `'`

AsherJingkongChen commented on build.js:
```diff
@@ -40,8 +63,34 @@
+
+    // Make output match what prettier expects
+    output = output.replaceAll('"', "'");
+    output = output.replace("'\n\]", "',\n]");
```
It just works.

hildjj replied:
Would you prefer to let prettier do this?  I can remove these lines.

AsherJingkongChen replied:
Since the code snippet should have a meaningful change for `index.json`, prettier can happily format the files.

AsherJingkongChen commented on build.js:
```diff
@@ -40,8 +63,34 @@
+    const hex = hash.digest('hex');
+
+    if (hex === previous?.hash) {
+      // There was a change to authors, etc. that we don't care about.
```
Wait, is author information contained in `output`?

AsherJingkongChen replied:
Wait, I understand it now. The snippet only executes if we successfully fetched the XML.

hildjj replied:
Author information is in the input XML file.  Once the XML is processed, you would get the same exact list of mediatypes as the previous run, even though fetch returned 304.

AsherJingkongChen commented:
Hi, I have done reviewing the codes, but there are some comments that requires your thoughts.

hildjj committed:
Escape lastModified, and let prettier do the formatting.

AsherJingkongChen committed:
(adjust): 3

1. Throws error instead of just exiting the process
2. Escape all the interpolated values
3. Reducing the lines

AsherJingkongChen committed:
Merge branch 'review/1' into fix-1

AsherJingkongChen commented:
@hildjj Hi, your changes look great! I just made a few minor tweaks to ensure everything is LGTM—hope you don’t mind.

hildjj commented on build.js:
```diff
@@ -22,20 +21,17 @@ try {
     if (previous?.lastModified) {
       init.headers = new Headers({
-        'if-modified-since': previous.lastModified,
+        'if-modified-since': previous?.lastModified,
```
No need for `?` here since we just checked it in the if statement.

AsherJingkongChen replied:
Right.

hildjj commented on build.js:
```diff
@@ -22,20 +21,17 @@ try {
-    if (!ok) {
-      throw new Error(`HTTP ${status} ${statusText}`);
+    const statusTextPretty = `HTTP ${status} ${statusText}`;
+    if (!ok || status === 304) {
```
This is incorrect in my opinion.  With this approach, you will retry 5 times with a 304 response.  Note that ok=false with a 304.

AsherJingkongChen replied:
Right.

hildjj commented on build.js:
```diff
@@ -63,37 +59,32 @@ try {
+    const lastModified = response.headers.get('last-modified') || null;
+    const updated = new Date().toUTCString();
+    if (hash === previous?.hash) {
+      throw new Error('No relevant changes detected.');
```
Same issue with the 304.  This will cause unnecessary retries.

AsherJingkongChen replied:
Yes.

hildjj reviewed (commented):
The change to exceptions from process.exit has caused a few issues, I think.

AsherJingkongChen committed:
(fix): Missing early returns

AsherJingkongChen commented:
> The change to exceptions from process.exit has caused a few issues, I think.

@hildjj Now the program early exits at the first try (if HTTP 304 and no relevant changes).

AsherJingkongChen committed:
(fix): It should check HTTP status code.

hildjj commented on build.js:
```diff
@@ -21,17 +21,16 @@ try {
     if (previous?.lastModified) {
       init.headers = new Headers({
-        'if-modified-since': previous?.lastModified,
+        'if-modified-since': previous.lastModified.toUTCString(),
```
It's safer to use the EXACT string returned from the server, since doing date parsing and toJSON is quite likely to introduce changes.

hildjj replied:
That would mean storing the date in index.js as a string, NOT as a date, as in my original.

AsherJingkongChen replied:
yes

hildjj reviewed (commented):
Suggest storing last-modified as a string.

AsherJingkongChen committed:
(adjust): Store dates as string instead of JS object

hildjj commented:
LGTM.  Thanks!
